Add `DISK_IOCTL_CTRL_DEINIT` and `DISK_IOCTL_CTRL_INIT` macros to de-initialize and initialize a disk device

danieldegrasse opened the pull request:
## Introduction

This is relatively simple change to the disk access API- it introduces two new IOCTL command codes:
- `DISK_IOCTL_CTRL_INIT`: initialize the disk, functionally equivalent to calling `disk_access_init`
- `DISK_IOCTL_CTRL_DEINIT`: de-initialize the disk. After this point the disk can be reinitialized via `disk_access_ioctl(disk, DISK_IOCTL_CTRL_INIT, NULL)` (or via `disk_access_init`)

### Problem description

Currently there is no way to explicitly de-initialize a disk device. This causes issues for devices like SD cards, which are inherently hot-pluggable. Moreover, some applications may wish to power off the SD card device when not using it (which would require first de-initializing it). Providing a method to do so via the disk access API allows users to realize additional power savings, or safely remove a hot-pluggable device such as an SD card from their board

### Dependencies

In order to make use of this API effectively, `fs_unmount` implementations for filesystems must be updated to take advantage of the new IOCTL calls. I have already made these changes for the FATFS filesystem implementation within this PR, but other filesystem implementations will need to be changed in order to use this feature

### Concerns and Unresolved Questions

The introduction of `DISK_IOCTL_CTRL_INIT` makes the `disk_access_init` API call redundant. However, given that the disk access API is stable I think that this redundancy does not clear the bar for a stable API change. If we'd like to do so, the primary benefit would be one less field in the API structure for each disk driver.

The other concern I have is if we need to provide a method for users to explicitly request a disk *not* be unmounted when calling `fs_unmount`. I think it is ok if `fs_unmount` implicitly de-initializes a disk device, since from what I can tell we do not support multiple partitions being present on one disk device (so if a filesystem is unmounted the disk should be de-initialized as well). However, this may not always be the case in the future.

danieldegrasse committed:
drivers: disk: add DISK_IOCTL_CTRL_DEINIT command to supported IOCTLs

Add DISK_IOCTL_CTRL_DEINIT ioctl command to disk subsystem. When
disk_access_ioctl() is called with this command, the disk will be
de-initialized. After this IOCTL completes, the disk can safely be
reinitialized.

Fixes #60628

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
fs: fat_fs: make IOCTL call to de-initialize disk in fatfs_unmount()

Make call to de-initialize disk in fatfs_unmount(). This will permit the
disk to be reinitialized when it is mounted with fatfs_mount().

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: fs: unmount and remount filesystem during init

Unmount and remount filesystem during init, to verify that filesystem
remount works as expected.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

de-nordic commented on subsys/fs/fat_fs.c:
```diff
@@ -463,9 +481,17 @@ static int fatfs_mount(struct fs_mount_t *mountp)
 
+	/* Make direct disk IOCTL call to deinit disk */
+	disk_res = disk_ioctl(((FATFS *)mountp->fs_data)->pdrv, CTRL_POWER, &param);
+	if (disk_res != RES_OK) {
+		return translate_disk_error(disk_res);
+	}
+
```
There should be some logging here. Otherwise how do you know whether this one failed or the one below if they now share common errno space?

danieldegrasse replied:
Good catch- I'll add this on next push, once we've determined how to enforce the requirements around INIT/DEINIT IOCTLS

de-nordic commented on drivers/disk/flashdisk.c:
```diff
@@ -405,6 +406,8 @@ static int disk_flash_access_ioctl(struct disk_info *disk, uint8_t cmd, void *bu
 		*(uint32_t *)buff = ctx->page_size / ctx->sector_size;
 		k_mutex_unlock(&ctx->lock);
 		return 0;
+	case DISK_IOCTL_CTRL_INIT:
+		return disk_flash_access_init(disk);
```
Shouldn't we follow this with a hard requirement that all operations on disk should return error on uninitialized device, unless the `DISK_IOCTL_CTRL_INIT` is issued?
Shouldn't `DISK_IOCTL_CTRL_DEINIT` make devices refuse to do anything unless init is issued again?

We can also go soft on requirement and just state that init is always required before we do anything an deinit will require additional init, with no check in code. That will require less code, but can it cause some strange errors?

de-nordic replied:
Lets go with the soft requirement. 
If that keeps biting us, we can change to configurable/hard requirements, to allow detection of problems.
But I guess that most such problems would be with things in a tree anyway, so on us.

The changes, if we choose to do them, would be limited to https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/disk/disk_access.c anyway, so no backends should require change if we do that later.

Hard requirement adds code and, I assume, will rarely trigger any error.

Do you agree on such assumptions?

danieldegrasse replied:
Yes, I do- let's make this a soft requirement

danieldegrasse committed:
doc: services: disk_access: add documentation for INIT/DEINIT IOCTLs

Add documentation for disk INIT/DEINIT IOCTLs to the disk access API.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

Laczen commented:
Hi @danieldegrasse I agree with your concern regarding future problems with partitions support. I think the logic of a unmount resulting in a disk deinit is not the best way. A different approach could be:
1. Add a pointer to the fs node (or a fs lock that could be added) as part of a disk structure,
2. When a fs is mounted this link is updated (registered),
3. When a disk de-init is needed, the fs node is used with the container_of macro to try an unmount the fs, any error is propagated to the de-init (e.g. trying to deinit when file is open),
4. When a fs is unmounted is done the link is updated to NULL (unregister),

For future partitions support the disk structure could be expanded with a partitions list (list of disk nodes) and each of these partitions could follow the above logic.

danieldegrasse commented:
> Hi @danieldegrasse I agree with your concern regarding future problems with partitions support. I think the logic of a unmount resulting in a disk deinit is not the best way. A different approach could be:
> 
>     1. Add a pointer to the fs node (or a fs lock that could be added) as part of a disk structure,
> 
>     2. When a fs is mounted this link is updated (registered),
> 
>     3. When a disk de-init is needed, the fs node is used with the container_of macro to try an unmount the fs, any error is propagated to the de-init (e.g. trying to deinit when file is open),
> 
>     4. When a fs is unmounted is done the link is updated to NULL (unregister),
> 
> 
> For future partitions support the disk structure could be expanded with a partitions list (list of disk nodes) and each of these partitions could follow the above logic.

Thanks for the feedback here. I'm not sure we need to make the filesystem structure a member of the disk structure, as this would involve some API changes I believe.

However, we could easily have a reference counter implemented on each disk. The reference counter would increment when the disk was initialized, and decrement when it was de-initialized. The actual "init" and "deinit" API calls would only be passed to the disk when the reference counter incremented from zero to one, or decremented from one to zero.

A future "disk partition" driver would pass each init call through to the disk device it was implementing a partition for. This way, the backing disk device would effectively reference count using the partitions. Does this seem like a good solution?

Laczen commented:
> However, we could easily have a reference counter implemented on each disk. The reference counter would increment when the disk was initialized, and decrement when it was de-initialized. The actual "init" and "deinit" API calls would only be passed to the disk when the reference counter incremented from zero to one, or decremented from one to zero.
> 
> A future "disk partition" driver would pass each init call through to the disk device it was implementing a partition for. This way, the backing disk device would effectively reference count using the partitions. Does this seem like a good solution?

Yes, this is a good solution. It wouldn't allow disk "deinit" to be driving the unmounting, but this might not be required. Adding a reference count is perfect for checking if the disk is in use (either by a fs or a partition) and to postpone the disk deinit until there are no more users.

danieldegrasse commented:
> Yes, this is a good solution. It wouldn't allow disk "deinit" to be driving the unmounting, but this might not be required. Adding a reference count is perfect for checking if the disk is in use (either by a fs or a partition) and to postpone the disk deinit until there are no more users.

The flow I envision is as follows:
`fs_mount` calls `disk_access_ioctl(DISK_IOCTL_CTRL_INIT)`, which increments disk reference count, and calls disk IOCTL operation if and only if disk reference count is moving from 0 to 1.
`fs_unmount` calls  `disk_access_ioctl(DISK_IOCTL_CTRL_DEINIT)`, which decrements disk reference count, and calls disk IOCTL operation if and only if disk reference count is moving from 1 to 0.

This way, the disk will still be deinitialized with `fs_unmount` in the basic (single disk, single filesystem) usecase. A "partitions" disk driver would instead simply call the `disk_access_ioctl(DISK_IOCTL_CTRL_DEINIT)` on its parent driver, which would only de-initialize the driver if there were no remaining users

Laczen reviewed (dismissed):
The reference counter for disk "usage" can be taken up in a different PR.

danieldegrasse commented:
> The reference counter for disk "usage" can be taken up in a different PR.

Unless there are any objections, I'm going to implement in this PR (hopefully sometime this week)- There will be a few subtle documentation changes needed to capture the reference counting concept, and I'd prefer to avoid an API change later on when those are implemented

de-nordic commented:
> > The reference counter for disk "usage" can be taken up in a different PR.
> 
> Unless there are any objections, I'm going to implement in this PR (hopefully sometime this week)- There will be a few subtle documentation changes needed to capture the reference counting concept, and I'd prefer to avoid an API change later on when those are implemented

I have no objections, although prepare for significant work with mount/umount because currently the file system drivers handle device type, so disk operations, not the mount/umount.

danieldegrasse committed:
disk_access: reference count initialization calls for disks

Reference count initialization calls for disks. This changes the
behavior of the disk_access_init() function, such that disks will no
longer be initialized again if the first disk access init call
succeeds.

Disk access is reference counted in preparation for supporting disk
de-initialization, where a balanced number of disk de-initialization
calls with disk initialization calls will de-initialize the disk.

Also, remove code in disk drivers that was already checking against
duplicate disk_access_init() calls.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: disk: add DISK_IOCTL_CTRL_INIT macro to initialize a disk

Add DISK_IOCTL_CTRL_INIT IOCTL to initialize a disk. This IOCTL is
intended to replace disk_access_init() for new applications, but
disk_access_init() is kept for legacy compatibility. The INIT IOCTL is
added to better match the path that will be used for disk
de-initialization. Like the disk_access_init() calls,
DISK_IOCTL_CTRL_INIT calls are reference counted

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
> I have no objections, although prepare for significant work with mount/umount because currently the file system drivers handle device type, so disk operations, not the mount/umount.

Made the update- the implementation I am using now essentially reference counts the INIT/DEINIT operations at the disk level, much like regulators handle enable/disable. This way, a balanced set of INIT/DEINIT calls to a disk will de initialize it.

As another advantage, disk drivers no longer need logic in their `init` implementations to prevent re initializing on duplicate calls

Laczen reviewed (dismissed):
LGTM

de-nordic commented on subsys/disk/disk_access.c:
```diff
@@ -58,9 +58,19 @@ int disk_access_init(const char *pdrv)
+				disk->refcnt++;
+			}
+		}
+	} else if ((disk != NULL) && (disk->refcnt < UINT16_MAX)) {
```
I know that possibility that refcnt gets to `UINT16_MAX` is impossible, but what is the point of just stopping increment when we get to `disk->refcnt` without message? I think that when we have reached that point there should be error issued, because something seems to call init for no reason in loop.

danieldegrasse replied:
We would issue an error, right? both the `if` and `else if` cases would be false, so we would return `-EINVAL`. Or do you think we should also add a logging statement

de-nordic replied:
I mean logging. Does this make sense?

danieldegrasse replied:
Sure, I have no issue adding that

de-nordic commented on drivers/disk/nvme/nvme_disk.c:
```diff
@@ -194,6 +193,11 @@ static int nvme_disk_ioctl(struct disk_info *disk, uint8_t cmd, void *buff)
 	return ret;
 }
 
+static int nvme_disk_init(struct disk_info *disk)
+{
+	return nvme_disk_ioctl(disk, DISK_IOCTL_CTRL_INIT, NULL);
+}
+
```
Since we end up calling `ioctl` with `DISK_IOCTL_CTRL_INIT`, should we consider dropping the `init` callback and replace contents of `disk_access_init` with call to the IOCTL, seems that they have the same code and we can save on some callbacks.

danieldegrasse replied:
I thought about this- the reason I don't want to drop `init` (yet) is because this would be a stable API change in that case. The documentation added in this PR already guides users to use the IOCTLs over `disk_access_init`, in the future we can (and should) deprecate `disk_access_init`. I just want to avoid doing it now to give users a transition window.

Do you think it would be best to mark `disk_access_init` as deprecated in this PR, and start the stable API change process? We can then remove `disk_access_init` in the next release cycle

de-nordic replied:
I do not mean in this pr, I mean should we log issue planning to do that.

danieldegrasse commented on subsys/fs/fat_fs.c:
```diff
@@ -463,9 +483,18 @@ static int fatfs_mount(struct fs_mount_t *mountp)
 	res = f_mount(NULL, translate_path(mountp->mnt_point), 0);
 
+	/* Make direct disk IOCTL call to deinit disk */
+	disk_res = disk_ioctl(((FATFS *)mountp->fs_data)->pdrv, CTRL_POWER, &param);
```
I agree we should probably not unmount the disk if `f_mount` fails. However, I'm not sure what you mean by allowing the user to force the operation? Perhaps we require them to issue the IOCTL directly in this case, like so:

```
res = fs_unmount(&fatfs_fs);
if (res < 0) {
    printf("Forcing unmount\n");
    disk_access_ioctl(fatfs_disk, DISK_IOCTL_CTRL_DEINIT, NULL);
}
```

Or do you mean we should include a "force" parameter with the `DISK_IOCTL_CTRL_DEINIT` IOCTL?

de-nordic replied:
You have now two power options ON and OFF. If unmount fails I am not sure whether we should power down and just return error, I think that the user now has to step in and do something.
If user does not care, for some reason, it is on the user to either repeat umount and issue powrdown or just powerdown the disk (which reminds me that `fs_unmount` should also probably have force argument).
Situation I image is where user wants to remove card from some device and does eject in phone app that connects to device, the device fails to unmount so now user gets message on screen, looks on logs and does the eject once again but this time ticks the "force on error" choice and continues - risk is on user.

danieldegrasse replied:
Sure, I agree that permitting a `force` argument makes sense. In the future, adding support for this at the `fs_unmount` level would make sense. For how, I will document that if a parameter is supplied to `DISK_IOCTL_CTRL_DEINIT`, that indicates the disk should be forcibly unmounted.

Something like this:

```
/** Deinitialize the disk. This IOCTL can be used to de-initialize the disk,
 * enabling it to be removed from the system if the disk is hot-pluggable.
 * Disk usage is reference counted, so for a given disk the
 * `DISK_IOCTL_CTRL_DEINIT` IOCTL must be issued as many times as the
 * `DISK_IOCTL_CTRL_INIT` IOCTL was issued in order to de-initialize it.
 * 
 * This macro optionally accepts a pointer to a boolean as the `buf` parameter,
 * which if true indicates the disk should be forcibly stopped, ignoring all reference
 * counts. The disk driver must report success if a forced stop is requested, but
 * this operation is inherently unsafe.
 */
```

As far as the `disk_ioctl` call in the fatfs unmount API, I think we should not force an unmount here, and should rely on the user to make an IOCTL directly for that case.

danieldegrasse commented:
reassigning to @decsny, as I continue to be unable to review my own PRs

fabiobaltieri commented:
@danieldegrasse I presume this is ok to be merged? Maybe drop the RFC tag from the title

danieldegrasse commented:
> @danieldegrasse I presume this is ok to be merged? Maybe drop the RFC tag from the title

Will do. Since it is a backwards compatible addition to a stable API (and I think that the relevant people have reviewed it) not planning to bring this to arch WG. So I think it is ok to merge
